### ISSUE TITLE
Storing Zero timestamp is not supported or not documented

### ISSUE THREAD

**Issue by Code7R** [mysql-net/MySqlConnector#501]
Hi, I am migrating from the Oracle version and there are a few issues, the most disturbing one is dealing with TIMESTAMP fields having a the special “zero” value aka '0000-00-00 00:00:00' - see https://dev.mysql.com/doc/refman/8.0/en/datetime.html and https://dev.mysql.com/doc/connector-net/en/connector-net-programming-datetime-invalid.html .

IMHO this is legal with Connector/NET and it worked well. However this was probably handled with the special MySqlDateTime type before and that code I had to change since MySqlConnector doesn't support this type, and I also had to set ConvertZeroDateTime to work around the representation issues on reading.

Anyhow, no matter what I assign as row item (DateTime.MinValue or even plain string like above), I end up with an exception about invalid datetime format on commit.

It would be nice have a way to pass a such special invalid date down to mysql (esp. as TIMESTAMP value).

**Comment by bgrainger** [comment]
I see at [this page](https://dev.mysql.com/doc/refman/8.0/en/datetime.html) that:

> MySQL does not accept `TIMESTAMP` values that include a zero in the day or month column or values that are not a valid date. The sole exception to this rule is the special “zero” value `'0000-00-00 00:00:00'`.

However, I'm not able to use that "special zero value" in STRICT mode (on by default in MySQL 5.7 and later) that includes the `NO_ZERO_DATE` [SQL mode](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_zero_date).

Are you running with `NO_ZERO_DATE` off? What server version and configuration do you have that allows `'0000-00-00 00:00:00'` as a valid value for a `TIMESTAMP` field?

**Comment by bgrainger** [comment]
OK. I think you will be seeing this in your server logs:

```
[Warning] 'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode. They will be merged with strict mode in a future release.
```

However, they haven't yet merged `NO_ZERO_DATE` into "strict mode" in MySQL 8.0 AFAICT.

(FWIW, the default MySQL SQL mode should have included `NO_ZERO_DATE` and `NO_ZERO_IN_DATE` since 5.7.8, so I'm not sure why they're off for you; however, that should give me enough information to try to reproduce the problem.)

**Comment by bgrainger** [comment]
> or even plain string like above

The following code works fine for me to insert a "zero" timestamp:

```
using (var cmd = new MySqlCommand("drop table if exists test.time; create table test.time(ts timestamp not null)", connection))
	cmd.ExecuteNonQuery();

using (var cmd = new MySqlCommand("insert into test.time values(@t);", connection))
{
	cmd.Parameters.AddWithValue("@t", "0000-00-00 00:00:00");
	cmd.ExecuteNonQuery();
}
```

Can you provide a code sample that demonstrates the problem, as well as the full exception message and callstack?

**Comment by Code7R** [comment]
Thanks, I rechecked the code and found another culprit but it did not solve the problem.
Some background: we are using DataSet and SqlDataAdapter MySqlDataAdapter to manage database IO.

The actual exception is now thrown when the string is being assigned to the DataRow item, like:

   ```
var drNewRow = _m_dataTable.NewRow();
   …

   object targetTypeVal = dateTimeObject;

   // some other logics, detecting default(DateTime) resp. DateTime.MinValue
   and assigning the string “0000-00-00 00:00:00” to targetTypeVal

   drNewRow[colName] = targetTypeVal;
```
And here we get System.ArgumentException, at:
```
     at System.Data.DataColumn.set_Item(Int32 record, Object value)

      at System.Data.DataRow.set_Item(DataColumn column, Object value)

      at System.Data.DataRow.set_Item(String columnName, Object value)
```

**Comment by bgrainger** [comment]
As far as I can tell, the exception is being thrown from the (Microsoft) `DataTable` code. This seems to happen because MySQL Server (without `NO_ZERO_DATE`) allows `0000-00-00 00:00:00` to be stored, which isn't a valid `DateTime` value. (As the exception message says, `String was not recognized as a valid DateTime.Couldn't store <0000-00-00 00:00:00> in ts Column. Expected type is DateTime.`)

This reproduces for me with Connector/NET, too. Do you have a code sample that works with Connector/NET but not with MySqlConnector, which would demonstrate a compatibility bug?

**Comment by bgrainger** [comment]
@Code7R can you please update to [v0.42.0](https://www.nuget.org/packages/MySqlConnector/0.42.0) to see if this fixes your compatibility problems? That version adds the `AllowZeroDateTime` connection string option and allows you to use `MySqlDateTime` to represent the zero datetime.

**Comment by bgrainger** [comment]
This should be fixed in 0.42.0. Closing due to no response; can reopen if there's still an issue.
